File: src/papernestextension/controls/material/papernest_color_picker.py
import json
import re
from typing import Optional

from flet.controls.base_control import control
from flet.controls.border_radius import BorderRadiusValue
from flet.controls.control_event import ControlEventHandler
from flet.controls.layout_control import LayoutControl
from flet.controls.material.form_field_control import InputBorder
from flet.controls.padding import PaddingValue
from flet.controls.text_style import TextStyle
from flet.controls.types import (
    ColorValue,
    IconDataOrControl,
    Number,
    StrOrControl,
    TextAlign,
)

__all__ = ["PaperNestColorPicker"]


@control("PaperNestColorPicker")
class PaperNestColorPicker(LayoutControl):
    """Sélecteur de couleur autonome à l'apparence d'un champ Material.

    Toute la surface du contrôle ouvre un dialogue contenant un MaterialPicker.
    La valeur publique est toujours normalisée au format ``#RRGGBB``.
    """

    value: Optional[ColorValue] = None

    picker_title: str = "Sélectionner une couleur"
    cancel_text: str = "Annuler"
    confirm_text: str = "Valider"
    enable_label: bool = False
    portrait_only: bool = False
    barrier_color: Optional[ColorValue] = None

    autofocus: bool = False
    read_only: bool = False
    text_align: TextAlign = TextAlign.START

    label: Optional[StrOrControl] = None
    hint_text: Optional[str] = "Sélectionner une couleur"
    helper: Optional[StrOrControl] = None
    helper_text: Optional[str] = None
    helper_max_lines: Optional[int] = None
    error: Optional[StrOrControl] = None
    error_text: Optional[str] = None
    error_max_lines: Optional[int] = None
    text_size: Optional[Number] = None
    text_style: Optional[TextStyle] = None
    label_style: Optional[TextStyle] = None
    floating_label_style: Optional[TextStyle] = None
    hint_style: Optional[TextStyle] = None
    hint_max_lines: Optional[int] = None
    helper_style: Optional[TextStyle] = None
    error_style: Optional[TextStyle] = None
    prefix: Optional[StrOrControl] = None
    prefix_style: Optional[TextStyle] = None
    prefix_icon: Optional[IconDataOrControl] = None
    clear_button: bool = False
    clear_icon: Optional[IconDataOrControl] = None
    clear_tooltip: str = "Effacer"
    border: Optional[InputBorder] = None
    color: Optional[ColorValue] = None
    focused_color: Optional[ColorValue] = None
    bgcolor: Optional[ColorValue] = None
    focused_bgcolor: Optional[ColorValue] = None
    focus_color: Optional[ColorValue] = None
    border_width: Number = 1
    border_color: Optional[ColorValue] = None
    border_radius: Optional[BorderRadiusValue] = None
    focused_border_width: Optional[Number] = None
    focused_border_color: Optional[ColorValue] = None
    content_padding: Optional[PaddingValue] = None
    dense: bool = False
    collapsed: bool = False
    align_label_with_hint: bool = False
    filled: bool = False
    fill_color: Optional[ColorValue] = None
    hover_color: Optional[ColorValue] = None

    on_change: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    on_clear: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    # Compatibilité avec les premières versions de l'extension.
    on_cleared: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    on_focus: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    on_blur: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    on_tap_outside: Optional[ControlEventHandler["PaperNestColorPicker"]] = None
    on_escape: Optional[ControlEventHandler["PaperNestColorPicker"]] = None

    @staticmethod
    def _normalize_hex(value) -> Optional[str]:
        if value is None:
            return None

        text = str(value).strip()
        if not text or text.lower() == "none":
            return None

        if not text.startswith("#"):
            return text

        digits = text[1:]
        if re.fullmatch(r"[0-9a-fA-F]{3}", digits):
            digits = "".join(char * 2 for char in digits)
        elif re.fullmatch(r"[0-9a-fA-F]{8}", digits):
            # Les couleurs Flet peuvent arriver en #AARRGGBB : l'API publique
            # de PaperNestColorPicker ne conserve volontairement que le RGB.
            digits = digits[2:]
        elif not re.fullmatch(r"[0-9a-fA-F]{6}", digits):
            return text

        return f"#{digits.upper()}"

    @classmethod
    def _color_from_event_data(cls, event_data) -> Optional[str]:
        if event_data is None:
            return None

        value = event_data
        if isinstance(value, dict):
            value = value.get("value") or value.get("color")
        elif isinstance(value, str):
            text = value.strip()
            if text.startswith("{"):
                try:
                    payload = json.loads(text)
                    value = payload.get("value") or payload.get("color") or text
                except (json.JSONDecodeError, TypeError):
                    value = text

        return cls._normalize_hex(value)

    def before_update(self):
        super().before_update()
        self.value = self._normalize_hex(self.value)
        if (
            self.bgcolor is not None
            or self.fill_color is not None
            or self.focused_bgcolor is not None
        ) and not self.filled:
            self.filled = True

    async def _trigger_event(self, event_name, event_data):
        if event_name == "change":
            self.value = self._color_from_event_data(event_data)
        elif event_name in ("clear", "cleared"):
            self.value = None
        await super()._trigger_event(event_name, event_data)

    async def focus(self) -> None:
        await self._invoke_method("focus")

    async def open(self) -> None:
        await self._invoke_method("open")

    async def clear(self) -> None:
        await self._invoke_method("clear")

File: src/flutter/papernestextension/lib/src/controls/papernest_color_picker.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';

import '../widgets/papernest_dialog_surface.dart';

class PaperNestColorPickerControl extends StatefulWidget {
  final Control control;

  PaperNestColorPickerControl({Key? key, required this.control})
      : super(key: key ?? ValueKey("control_${control.id}"));

  @override
  State<PaperNestColorPickerControl> createState() =>
      _PaperNestColorPickerControlState();
}

class _PaperNestColorPickerControlState
    extends State<PaperNestColorPickerControl> {
  late final FocusNode _focusNode;
  Color? _value;
  bool _focused = false;
  bool _hovered = false;

  bool get _readOnly => widget.control.getBool("read_only", false)!;
  bool get _interactive => !widget.control.disabled && !_readOnly;

  @override
  void initState() {
    super.initState();
    _focusNode = FocusNode();
    _focusNode.addListener(_onFocusChange);
    widget.control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    _focusNode.removeListener(_onFocusChange);
    _focusNode.dispose();
    widget.control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }

  void _onFocusChange() {
    if (_focused == _focusNode.hasFocus) return;
    setState(() => _focused = _focusNode.hasFocus);
    widget.control.triggerEvent(_focused ? "focus" : "blur");
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    switch (name) {
      case "focus":
        _focusNode.requestFocus();
        return null;
      case "open":
        await _openPicker();
        return null;
      case "clear":
        _clearValue();
        return null;
      default:
        throw Exception("Unknown PaperNestColorPicker method: $name");
    }
  }

  Color? _backendColor(BuildContext context) {
    final raw = widget.control.get("value");
    if (raw == null) return null;
    final text = raw.toString().trim();
    if (text.isEmpty || text.toLowerCase() == "none") return null;
    return widget.control.getColor("value", context);
  }

  String _colorHex(Color color) {
    final rgb = color.value & 0x00ffffff;
    return '#${rgb.toRadixString(16).padLeft(6, '0').toUpperCase()}';
  }

  Future<void> _openPicker() async {
    if (!_interactive) return;
    if (!_focusNode.hasFocus) _focusNode.requestFocus();

    var temporaryColor = _value ?? Theme.of(context).colorScheme.primary;

    final selected = await showDialog<Color>(
      context: context,
      barrierColor: widget.control.getColor("barrier_color", context),
      builder: (dialogContext) {
        return StatefulBuilder(
          builder: (context, setDialogState) {
            return PaperNestDialogSurface(
              control: widget.control,
              variant: "primary",
              width: widget.control.getDouble("dialog_width", 560),
              maxHeight: widget.control.getDouble("dialog_max_height", 720),
              title: Text(
                widget.control.getString(
                  "picker_title",
                  "Sélectionner une couleur",
                )!,
              ),
              icon: const Icon(Icons.palette_outlined),
              content: MaterialPicker(
                pickerColor: temporaryColor,
                onColorChanged: (color) {
                  setDialogState(() => temporaryColor = color);
                },
                enableLabel:
                    widget.control.getBool("enable_label", false)!,
                portraitOnly:
                    widget.control.getBool("portrait_only", false)!,
              ),
              actions: [
                TextButton(
                  onPressed: () => Navigator.of(dialogContext).pop(),
                  child: Text(
                    widget.control.getString("cancel_text", "Annuler")!,
                  ),
                ),
                FilledButton.icon(
                  onPressed: () =>
                      Navigator.of(dialogContext).pop(temporaryColor),
                  icon: const Icon(Icons.check_rounded),
                  label: Text(
                    widget.control.getString("confirm_text", "Valider")!,
                  ),
                ),
              ],
            );
          },
        );
      },
    );

    if (selected == null) return;
    setState(() => _value = selected);
    final hex = _colorHex(selected);
    widget.control.updateProperties({"value": hex});
    widget.control.triggerEvent("change", hex);
  }

  void _clearValue() {
    if (!_interactive || _value == null) return;
    setState(() => _value = null);
    widget.control.updateProperties({"value": null});
    widget.control.triggerEvent("clear");
    widget.control.triggerEvent("cleared");
    widget.control.triggerEvent("change", "");
    _focusNode.requestFocus();
  }

  void _handleEscape() {
    widget.control.triggerEvent("escape");
    if (_value != null) _clearValue();
  }

  InputBorder _border(BuildContext context, {required bool focused}) {
    final borderType = widget.control.getString("border") ?? "outline";
    if (borderType == "none") return InputBorder.none;

    final width = focused
        ? widget.control.getDouble("focused_border_width") ??
            widget.control.getDouble("border_width", 1)!
        : widget.control.getDouble("border_width", 1)!;
    final color = focused
        ? widget.control.getColor("focused_border_color", context) ??
            widget.control.getColor("border_color", context) ??
            Theme.of(context).colorScheme.primary
        : widget.control.getColor("border_color", context) ??
            Theme.of(context).colorScheme.outline;
    final side = width == 0
        ? BorderSide.none
        : BorderSide(color: color, width: width);

    if (borderType == "underline") {
      return UnderlineInputBorder(borderSide: side);
    }
    return OutlineInputBorder(
      borderSide: side,
      borderRadius: widget.control.getBorderRadius("border_radius") ??
          const BorderRadius.all(Radius.circular(4)),
    );
  }

  Widget _buildPrefixIcon() {
    return widget.control.buildIconOrWidget("prefix_icon") ??
        const Icon(Icons.color_lens_outlined);
  }

  Widget _buildValueContent(TextStyle textStyle) {
    if (_value == null) return const SizedBox.shrink();

    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        DecoratedBox(
          decoration: BoxDecoration(
            color: _value,
            shape: BoxShape.circle,
            border: Border.all(
              color: Theme.of(context).colorScheme.outlineVariant,
            ),
          ),
          child: const SizedBox(width: 18, height: 18),
        ),
        const SizedBox(width: 10),
        Text(
          _colorHex(_value!),
          style: textStyle,
          textAlign:
              widget.control.getTextAlign("text_align", TextAlign.start),
        ),
      ],
    );
  }

  Widget? _buildClearButton() {
    final clearButton = widget.control.getBool("clear_button", false)!;
    if (!clearButton || _value == null || !_interactive) return null;

    final icon = widget.control.buildIconOrWidget("clear_icon") ??
        const Icon(Icons.clear_rounded);
    return IconButton(
      icon: icon,
      tooltip: widget.control.getString("clear_tooltip", "Effacer"),
      onPressed: _clearValue,
    );
  }

  @override
  Widget build(BuildContext context) {
    final backendValue = _backendColor(context);
    if ((backendValue == null) != (_value == null) ||
        (backendValue != null &&
            _value != null &&
            backendValue.value != _value!.value)) {
      _value = backendValue;
    }

    final theme = Theme.of(context);
    final color = widget.control.getColor("color", context);
    final focusedColor = widget.control.getColor("focused_color", context);
    final textStyle = (widget.control.getTextStyle("text_style", theme) ??
            theme.textTheme.bodyLarge ??
            const TextStyle())
        .copyWith(
      fontSize: widget.control.getDouble("text_size"),
      color: _focused ? focusedColor ?? color : color,
    );

    final bgcolor = widget.control.getColor("bgcolor", context);
    final focusedBgcolor =
        widget.control.getColor("focused_bgcolor", context);
    final fillColor = widget.control.getColor("fill_color", context) ??
        (_focused ? focusedBgcolor ?? bgcolor : bgcolor);

    final decoration = InputDecoration(
      enabled: !widget.control.disabled,
      label: widget.control.buildTextOrWidget("label"),
      labelStyle: widget.control.getTextStyle("label_style", theme),
      floatingLabelStyle:
          widget.control.getTextStyle("floating_label_style", theme),
      hintText: widget.control.getString("hint_text"),
      hintStyle: widget.control.getTextStyle("hint_style", theme),
      hintMaxLines: widget.control.getInt("hint_max_lines"),
      helper: widget.control.buildTextOrWidget("helper"),
      helperText: widget.control.getString("helper_text"),
      helperStyle: widget.control.getTextStyle("helper_style", theme),
      helperMaxLines: widget.control.getInt("helper_max_lines"),
      error: widget.control.buildTextOrWidget("error"),
      errorText: widget.control.getString("error_text"),
      errorStyle: widget.control.getTextStyle("error_style", theme),
      errorMaxLines: widget.control.getInt("error_max_lines"),
      prefix: widget.control.buildTextOrWidget("prefix"),
      prefixStyle: widget.control.getTextStyle("prefix_style", theme),
      prefixIcon: _buildPrefixIcon(),
      prefixIconConstraints:
          widget.control.getBoxConstraints("prefix_icon_constraints"),
      suffixIcon: _buildClearButton(),
      suffixIconConstraints:
          widget.control.getBoxConstraints("suffix_icon_constraints"),
      isDense: widget.control.getBool("dense", false)!,
      isCollapsed: widget.control.getBool("collapsed", false)!,
      alignLabelWithHint:
          widget.control.getBool("align_label_with_hint", false),
      filled: widget.control.getBool("filled", fillColor != null)!,
      fillColor: fillColor,
      hoverColor: widget.control.getColor("hover_color", context),
      focusColor: widget.control.getColor("focus_color", context),
      contentPadding: widget.control.getPadding("content_padding"),
      constraints: widget.control.getBoxConstraints("size_constraints"),
      border: _border(context, focused: false),
      enabledBorder: _border(context, focused: false),
      focusedBorder: _border(context, focused: true),
      disabledBorder: _border(context, focused: false),
    );

    final field = FocusableActionDetector(
      focusNode: _focusNode,
      autofocus: widget.control.getBool("autofocus", false)!,
      enabled: !widget.control.disabled,
      mouseCursor: _interactive
          ? SystemMouseCursors.click
          : SystemMouseCursors.basic,
      onShowFocusHighlight: (value) {
        if (_focused != value && mounted) setState(() => _focused = value);
      },
      onShowHoverHighlight: (value) {
        if (_hovered != value && mounted) setState(() => _hovered = value);
      },
      shortcuts: const <ShortcutActivator, Intent>{
        SingleActivator(LogicalKeyboardKey.enter): ActivateIntent(),
        SingleActivator(LogicalKeyboardKey.space): ActivateIntent(),
        SingleActivator(LogicalKeyboardKey.escape): DismissIntent(),
      },
      actions: <Type, Action<Intent>>{
        ActivateIntent: CallbackAction<ActivateIntent>(
          onInvoke: (_) {
            _openPicker();
            return null;
          },
        ),
        DismissIntent: CallbackAction<DismissIntent>(
          onInvoke: (_) {
            _handleEscape();
            return null;
          },
        ),
      },
      child: Material(
        type: MaterialType.transparency,
        child: InkWell(
          mouseCursor: _interactive
              ? SystemMouseCursors.click
              : SystemMouseCursors.basic,
          canRequestFocus: false,
          onTap: _interactive ? _openPicker : null,
          hoverColor: widget.control.getColor("hover_color", context),
          focusColor: widget.control.getColor("focus_color", context),
          borderRadius: widget.control.getBorderRadius("border_radius"),
          child: InputDecorator(
            isFocused: _focused,
            isHovering: _hovered,
            isEmpty: _value == null,
            decoration: decoration,
            child: _buildValueContent(textStyle),
          ),
        ),
      ),
    );

    return LayoutControl(control: widget.control, child: field);
  }
}
